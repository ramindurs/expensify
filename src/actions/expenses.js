import {v4 as uuidv4} from "uuid";
import database from '../firebase/firebase';

const FB_REF = 'expenses';

export const addExpense = (expense) => ({
        type: 'ADD_EXPENSE',
        expense
    }
);

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = {
            description,
            note,
            amount,
            createdAt
        };
        return database().ref(FB_REF).push(expense).then((fbRef) => {
            dispatch(addExpense({
                id: fbRef.key,
                ...expense
            }))
        })
    };
};

export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = (id) => {
    return (dispatch) => {
        const ref = `${FB_REF}/${id}`;
        return database().ref(ref).remove().then(() => {
            return dispatch(removeExpense(id));
        }).catch((error) => {
            console.error('Error removing ', ref);
        });
    };
};

export const editExpense = ({id, updates}) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = ({id, updates}) => {
    return (dispatch) => {
        const ref = `${FB_REF}/${id}`;
        return database().ref(ref).update(updates).then(() => {
            return dispatch(editExpense({id, updates}));
        }).catch((error) => {
            console.error('Error updating', ref);
        });
    }
};

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return database().ref(FB_REF).once('value')
            .then((dataSnapshot) => {
                const expenses = [];
                dataSnapshot.forEach((childSnapshot) => {
                    expenses.push({
                        ...childSnapshot.val(),
                        id: childSnapshot.key
                    });
                });
                return dispatch(setExpenses(expenses));
            });
    };
};
