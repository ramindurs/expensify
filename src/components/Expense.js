import React from 'react';
import {Link} from 'react-router-dom';
import moment from "moment";
import numeral from 'numeral';

export const Expense = ({ id, description, amount, createdAt, dispatch, history}) => {
    return (
        <div>
            {id} <Link to={`/edit/${id}` }>{description}</Link>
            <p>
                {numeral(amount / 100).format('0,0.00')}
                -
                {moment(createdAt).format('MMM Do YYYY')}
            </p>
        </div>
    );
};

export default Expense;
