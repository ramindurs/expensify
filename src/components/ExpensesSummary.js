import React from 'react';
import {connect} from "react-redux";
import selectExpenses from '../selectors/expenses';
import expensesTotal from "../selectors/expenses-total";

export class ExpensesSummary extends React.Component {
    getTotalNumber = () => (this.props.expenses.length);
    getTotalAmount = () => (this.props.expensesTotal(this.props.expenses));

    render() {
        return (
            <div>
                <h4>Expenses Summary</h4>
                {this.getTotalNumber() === 0 ? 'Number of Expenses: 0' :
                    'Number of Expenses: ' + this.getTotalNumber() + ' Total Amount: Â£' + this.getTotalAmount()
                }
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters),
        expensesTotal: (filteredExpenses) => (expensesTotal(filteredExpenses) / 100)
    }
};

export default connect(mapStateToProps)(ExpensesSummary);
