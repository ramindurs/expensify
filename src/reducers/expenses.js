
const expensesRedecuerDefaultState = [];

export default (state = expensesRedecuerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter(ex => ex.id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id !== action.id ? expense : {
                ...expense,
                ...action.updates
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};