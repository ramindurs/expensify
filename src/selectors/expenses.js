import moment from "moment";

export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment) : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment) : true;
        const textMatch = expense.description.toUpperCase().indexOf(text.toUpperCase()) !== -1;
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => (sortBy === 'date' ? b.createdAt - a.createdAt: b.amount - a.amount));
};