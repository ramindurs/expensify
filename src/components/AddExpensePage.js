import React from 'react';
import {connect} from 'react-redux';
import {startAddExpense} from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/dashboard');
    };

    render() {
        return (
            <div>
                <div className='page-header-layout'>
                    <div className="content_container">
                        <h1 className='page-header-layout__title'>Add Expenses</h1>
                    </div>
                </div>
                <div className="content_container">
                    <ExpenseForm
                        submitMessage={'Save Expense'}
                        onSubmit={this.onSubmit}/>
                </div>
            </div>
        );
    }
}

const mapDispatchTpProps = (dispatch) => {
    return {
        startAddExpense: (expense) => dispatch(startAddExpense(expense))
    };
};

export default connect(undefined, mapDispatchTpProps)(AddExpensePage);
