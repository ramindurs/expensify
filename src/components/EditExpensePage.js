import React from 'react';
import {connect} from "react-redux";
import ExpenseForm from "./ExpenseForm";
import {startEditExpense, startRemoveExpense} from "../actions/expenses";

export class EditExpensePage extends React.Component {
    removeExpense = (id) => {
        this.props.startRemoveExpense(id);
        this.props.history.push('/dashboard');
    };
    editExpense = (id, expense) => {
        this.props.startEditExpense(id, expense);
        this.props.history.push('/dashboard');
    };

    render() {
        return (
            <div>
                <h3>Edit Expense ID {this.props.expense.id}</h3>
                <ExpenseForm
                    expenseToEdit={this.props.expense}
                    submitMessage={'Update Expense'}
                    onSubmit={(expense) => {
                        this.editExpense(this.props.expense.id, expense)
                    }}
                />
                <button onClick={() => {
                    this.removeExpense(this.props.expense.id)
                }
                }>Delete
                </button>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(e => e.id === props.match.params.id)
    };
};

const matchDispatchToProps = (dispatch, props) => {
    return {
        startEditExpense: (id, expense) => dispatch(startEditExpense({
            id: id,
            updates: expense
        })),
        startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
    };
};

export default connect(mapStateToProps, matchDispatchToProps)(EditExpensePage);
