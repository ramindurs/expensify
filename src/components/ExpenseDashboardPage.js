import React from 'react';
import Expenses from "./Expenses";
import ExpensesFilter from "./ExpensesFilter";
import ExpensesSummary from "./ExpensesSummary";

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesFilter/>
        <ExpensesSummary/>
        <Expenses/>
    </div>
);

export default ExpenseDashboardPage;