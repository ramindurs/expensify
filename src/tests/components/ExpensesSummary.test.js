import React from 'react';
import {shallow} from 'enzyme';
import {ExpensesSummary} from "../../components/ExpensesSummary";
import expenses from "../fixtures/expenses";

const expensesTotal = (ex) => (ex.map(e => e.amount).reduce((a, b) => a + b, 0) / 100);

test('should render expenses summary component', () => {
    const wrapper = shallow(<ExpensesSummary
        expenses={[]}
    />);

    expect(wrapper).toMatchSnapshot();
});

test('should render expenses when there is one expense', () => {
    const wrapper = shallow(<ExpensesSummary
        expenses={[expenses[0]]}
        expensesTotal={expensesTotal}
    />);

    expect(wrapper).toMatchSnapshot();
});

test('should render expenses when there are more than one expense', () => {
    const wrapper = shallow(<ExpensesSummary
        expenses={expenses}
        expensesTotal={expensesTotal}
    />);

    expect(wrapper).toMatchSnapshot();
});