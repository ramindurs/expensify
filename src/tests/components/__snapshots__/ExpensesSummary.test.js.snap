// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expenses summary component 1`] = `
<div>
  <h4>
    Expenses Summary
  </h4>
  Number of Expenses: 0
</div>
`;

exports[`should render expenses when there are more than one expense 1`] = `
<div>
  <h4>
    Expenses Summary
  </h4>
  Number of Expenses: 3 Total Amount: £245
</div>
`;

exports[`should render expenses when there is one expense 1`] = `
<div>
  <h4>
    Expenses Summary
  </h4>
  Number of Expenses: 1 Total Amount: £15
</div>
`;
